src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperContextExtensions.cs

```diff
@@ -28,8 +28,81 @@ public static class TagHelperContextExtensions
+                    logger.WriteWarning(new MissingAttributeLoggerStructure(context.UniqueId, missing));
+                }
+
+                return false;
+            }
+
+            // All required attributes present
+            return true;
+        }
+
+        /// <summary>
+        /// Determines the mode a <see cref="ITagHelper" /> should run in based on which mode has all its required
+        /// attributes present, non null, non empty, and non whitepsace.
+        /// </summary>
+        /// <typeparam name="TMode">The type representing the <see cref="ITagHelper" />'s modes.</typeparam>
+        /// <typeparam name="TSet">The type representing which attributes are required for which mode.</typeparam>
+        /// <param name="context">The <see cref="TagHelperContext"/>.</param>
+        /// <param name="modeRequiredAttributes">The modes and their required attributes.</param>
+        /// <param name="logger">An <see cref="ILogger"/> to log messages to.</param>
+        /// <returns>The <see cref="ModeMatchResult{TMode}"/>.</returns>
+        public static ModeMatchResult<TMode> DetermineMode<TMode, TSet>(
+            this TagHelperContext context,
+            IEnumerable<Tuple<TMode, TSet>> modeRequiredAttributes,
+            ILogger logger = null)
+            where TSet : IEnumerable<string>
+        {
+            List<Tuple<TMode, IEnumerable<string>>> modeCandidates = null;
+
+            foreach (var set in modeRequiredAttributes)
+            {
+                var attributes = GetPresentMissingAttributes(context, set.Item2);
+                var present = attributes.Item1;
+                var missing = attributes.Item2;
+
+                if (!missing.Any())
+                {
+                    return ModeMatchResult.Matched(set.Item1);
+                }
+
+                if (missing.Any() && present.Any())
+                {
+                    // The set had some present attributes but others missing so capture details of those missing to
+                    // log later on if no match is found
+                    modeCandidates = modeCandidates ?? new List<Tuple<TMode, IEnumerable<string>>>();
+                    modeCandidates.Add(Tuple.Create(set.Item1, missing));
+                }
+            }
+
+            // If a partial was match found, log a warning
+            if (modeCandidates != null && logger != null && logger.IsEnabled(LogLevel.Warning))
+            {
+                logger.WriteWarning(new PartialModeMatchLoggerStructure(context.UniqueId,
+                    modeCandidates.Select(candidate =>
+                        Tuple.Create(candidate.Item1.ToString(), candidate.Item2.ToArray()))));
+            }
+
+            return ModeMatchResult<TMode>.Unmatched;
+        }
+
+        private static Tuple<IEnumerable<string>, IEnumerable<string>> GetPresentMissingAttributes(
+            [NotNull]this TagHelperContext context,
```

Comment:
Make this a regular method. Also drop the `NotNull` - shouldnt need those on private methods.